landoapi/workers/base.py

```diff
@@ -0,0 +1,136 @@
+            # of the file.
+            self.ssh_private_key = os.environ.get(SSH_PRIVATE_KEY_ENV_KEY)
+            if not self.ssh_private_key:
+                logger.warning(f"No {SSH_PRIVATE_KEY_ENV_KEY} present in environment.")
+
+    @staticmethod
+    def _setup_ssh(ssh_private_key):
+        """Add a given private ssh key to ssh agent.
+
+        SSH keys are needed in order to push to repositories that have an ssh
+        push path.
+
+        The private key should be passed as it is in the key file, including all
+        new line characters and the new line character at the end.
+
+        Args:
+            ssh_private_key (str): A string representing the private SSH key file.
+        """
+        # Set all the correct environment variables
+        agent_process = subprocess.run(
+            ["ssh-agent", "-s"], capture_output=True, universal_newlines=True
+        )
+
+        # This pattern will match keys and values, and ignore everything after the
+        # semicolon. For example, the output of `agent_process` is of the form:
+        #     SSH_AUTH_SOCK=/tmp/ssh-c850kLXXOS5e/agent.120801; export SSH_AUTH_SOCK;
+        #     SSH_AGENT_PID=120802; export SSH_AGENT_PID;
+        #     echo Agent pid 120802;
+        pattern = re.compile("(.+)=([^;]*)")
+        for key, value in pattern.findall(agent_process.stdout):
+            logger.info(f"_setup_ssh: setting {key} to {value}")
+            os.environ[key] = value
+
+        # Add private SSH key to agent
+        # NOTE: ssh-add seems to output everything to stderr, including upon exit 0.
+        add_process = subprocess.run(
+            ["ssh-add", "-"],
+            input=ssh_private_key,
+            capture_output=True,
+            universal_newlines=True,
+        )
+        if add_process.returncode != 0:
+            raise Exception(add_process.stderr)
+        logger.info("Added private SSH key from environment.")
+
+    @property
+    def _paused(self):
+        return ConfigurationVariable.get(self.PAUSE_KEY, False)
+
+    @property
+    def _running(self):
+        # When STOP_KEY is False, the worker is running.
+        return not ConfigurationVariable.get(self.STOP_KEY, False)
+
+    def _setup(self):
+        if hasattr(self, "ssh_private_key"):
+            self._setup_ssh(self.ssh_private_key)
+
+    def _start(self, max_loops=None, *args, **kwargs):
+        loops = 0
+        while self._running:
+            if max_loops is not None and loops >= max_loops:
+                break
+            while self._paused:
+                self.throttle(self.sleep_seconds)
+            self.loop(*args, **kwargs)
+            loops += 1
+
+        logger.info(f"{self} exited after {loops} loops.")
+
+    @property
+    def throttle_seconds(self):
+        return ConfigurationVariable.get(self.THROTTLE_KEY, 3)
+
+    def throttle(self, seconds=None):
+        sleep(seconds if seconds is not None else self.throttle_seconds)
```

Comment:
`self.throttle_seconds` can technically be a `bool` - we should ensure it is a `float` before using it.

Reply:
How can it be a `bool`? The default is `3` (which is an int). It is not set in the DB (yet).

Once we deploy this, we can initialize the database value, and `variable_type` would have to be set to an `INT` on a one-time basis, but it makes no sense for it to be set to a `bool`. The type checking is the responsibility of whomever sets the initial values for these variables in the database (or if there is a command, like in `cli` which sets these values.) They are not designed to be changed arbitrarily.

Reply:
The returned type from `ConfigurationVariable.get` is `ConfigurationValueType = Union[bool, int, str]`, so the type checker detects this can possibly be a `bool`.

Perhaps instead of a `get` method for retrieving the variables we should have a `get_bool`, `get_int` etc.

Reply:
> The returned type from `ConfigurationVariable.get` is `ConfigurationValueType = Union[bool, int, str]`, so the type checker detects this can possibly be a `bool`.
> 
> Perhaps instead of a `get` method for retrieving the variables we should have a `get_bool`, `get_int` etc.

We don't have any enforcement of type checking (not sure if you have a way to set your IDE to ignore this particular case if it is bugging you). This would be out of scope for this PR, however I agree that there could be some improvements made to how the DB configuration variable system works. Perhaps you can file a bug for this, as I think it requires further discussion.

Reply:
We can discuss adding helper methods for getting variables of the correct type from the db.

For now we should just wrap the value in a `float` before we use it.

Reply:
> We can discuss adding helper methods for getting variables of the correct type from the db.
> 
> For now we should just wrap the value in a `float` before we use it.

But it's an `int`, not a `float`. And what difference would that make other than suppress the warning in your IDE? It will throw an error regardless if it gets the wrong input.

I would like to avoid adding code that will just help with IDE-related issues, unless it also has a more general benefit.

Adding a try/except on the other hand may accomplish something if you're trying to handle an incorrect type here.

Reply:
See: https://github.com/mozilla-conduit/lando-api/blob/main/landoapi/models/configuration.py#L64

Reply:
Probably worth noting that `time.sleep` does not care about the variable type it gets (and in fact would work if provided a bool, though it would interpret those as integers 0 and 1). I think [this](https://github.com/python/cpython/blob/main/Python/pytime.c#L553) is where it interprets the argument.

Reply:
I'll just fix it myself after this lands.